### PR TITLE
Support chunk variants

### PR DESCRIPTION
**Pull request by csweichel**
This PR adds support for chunk variants that vary build args and changes the naming of `--chunked-without-hash` builds.

## Chunk Variants
With this change, chunks can have variants that vary build args or the `Dockerfile` entirely. This feature is intended to provide e.g. multiple versions of the same language, e.g. Golang. Chunk variants are identified in a `name:tag` notation, e.g. if the `golang` chunk has two variants `1.15.0` and `1.16.0` a combination might refer to them as `golang:1.15.0` resp. `golang:1.16.0`.

To make a chunk variant, add a `chunk.yaml` much like this one:
```YAML
variants:
  - name: 1.15.0
    args:
        GO_VERSION: 1.15.0
```
which would introduce a single variant that varies the `GO_VERSION` build argument.

## Chunked without hash
Prior to this change, all chunks built with `--chunked-without-hash` would result in `target/repository:chunk-name` images. Now, such chunks result in `target/repository/chunk-name:variant` images. If a chunk doesn't have variants, we'll use `latest` as tag.

## How to test
```bash
# prepare
go generate -v ./...

# install buildkitd
sudo su -c "cd /usr; curl -L https://github.com/moby/buildkit/releases/download/v0.8.3/buildkit-v0.8.3.linux-amd64.tar.gz | tar xvz"

# start buildkit
sudo buildkitd --group gitpod

# start a registry (make sure port 5000 is public)
docker run -p 5000:5000 registry

# build the chunks
go run main.go build --context example $(gp url 5000 | cut -d/ -f3)/dazzle

```

### CODE REVIEW

File: example/chunks/node/Dockerfile [hunk 1]
```diff
@@ -1,5 +1,13 @@
+
 ARG base
 FROM ${base}
 
-RUN apk add --no-cache nodejs
-ENV LANG=en_US.UTF-8
\ No newline at end of file
+ARG NODE_VERSION
+
+RUN curl -fsSL https://raw.githubusercontent.com/nvm-sh/nvm/v0.37.2/install.sh | PROFILE=/dev/null bash
```
Comment (rl-gitpod) [inline example/chunks/node/Dockerfile:7]: I know it is an example but how are we intending to handle versioning/upgrading nvm going forward? An nvm chunk?

File: pkg/dazzle/project.go [hunk 2]
```diff
@@ -153,7 +166,7 @@ func LoadFromDir(dir string) (*Project, error) {
 
 	res := &Project{
 		Config: *cfg,
-		Base:   *base,
+		Base:   base[0],
```
Comment (rl-gitpod) [inline pkg/dazzle/project.go:169]: Might want to detect an empty variants?

File: pkg/dazzle/project.go [hunk 3]
```diff
@@ -171,38 +184,81 @@ func LoadFromDir(dir string) (*Project, error) {
 		if !chd.IsDir() {
 			continue
 		}
-		chnk, err := loadChunk(chunksDir, testsDir, chd.Name())
+		chunks, err := loadChunk(chunksDir, testsDir, chd.Name())
```
Comment (rl-gitpod) [inline pkg/dazzle/project.go:187]: /minor nit/
loadChunks to match?

**Review by rl-gitpod — APPROVED** [review]

File: example/chunks/node/Dockerfile (hunk 1 above)
Comment (csweichel) [reply example/chunks/node/Dockerfile]: Good point. I've introduced an `NVM_VERSION` arg that we could make variant if we wanted to

File: pkg/dazzle/project.go (hunk 2 above)
Comment (csweichel) [reply pkg/dazzle/project.go:172]: I've added a check to this effect

**Pull request merged**